### PR TITLE
Create mad6t_oneturn.py and study.py

### PR DESCRIPTION
**Pull request by luxiaohan**
Create new python script mad6t_oneturn.py for madx and one turn sixtrack job, create new module study.py to read variables from bash shell and generate input file for mad6t_oneturn.py

### CODE REVIEW

**Comment by amereghe** [comment]
Hello @luxiaohan,
thanks for the pull request. I am sorry that I am able to go through it only now. I won't be able to comment on everything in one go and right now, I will go through the code in steps. Please feel free to discuss here my remarks and, in case, apply the changes.
Thanks!

**Comment by amereghe** [comment]
Let's start with the easy part - I would change slightly the tree structure, i.e.
```
./
   |_ README.md
   |_ lib/
   |  |_ study.py
   |  |_ mad6t_oneturn.py
   |_ templates/
   |  |_ mad6t.ini
   |  |_ fort.3.mother1
   |  |_ fort.3.mother2
   |  |_ hl10.mask
   |_ sandbox/
      |_ run_mad6t.py
```

The `run_mad6t.py` is actually your `testspace/run/mad6t_oneturn.py`, plus a couple of command to copy locally files from template if they are not there.

**Comment by luxiaohan** [comment]
Hello @amereghe 
Ok, I will apply the changes of the tree structure. `testspace/run/mad6t_oneturn.py`  is a symbolic link of `mad6t_oneturn.py` for testing. Now mad6t_oneturn.py can read the required files from templates directly, do you mean we need to copy the files from templates to the current working directory at first?  e.g. copy `templates/fort.3.mother1` to `./`.

**Comment by amereghe** [comment]
Hi @luxiaohan ,
honestly, I would avoid using links, as it is better to have the code saying it all...
For the copy, yes, please, add a copy statement

**Comment by luxiaohan** [comment]
Hi @amereghe,
ok, I will modify `mad6t_oneturn.py`. And if we don't use link for testing, we will need to copy the script from `./lib` to `./sandbox` after every change. Or I can create a new script under `./sandbox` for testing, e.g.
 

> run_mad6t.py
`../lib/mad6t_oneturn.py  mad6t.ini`

**Comment by amereghe** [comment]
why not using `import`?

**Comment by luxiaohan** [comment]
Now, `mad6t_oneturn.py` is an executable script, it accepts arguments from system (`mad6t.ini`). If we use `import`, we need to change the way to execute `mad6t_oneturn.py`.

**Comment by amereghe** [comment]
I think we can have a script in the `sandbox` which takes care of copying (if needed) and run the `mad6t_oneturn.py`, can't we?

**Comment by luxiaohan** [comment]
Of course we can. As I said before, I can create a script `run_mad6t.py` in the `sandbox` to run the `mad6t_oneturn.py`. But where should we put the copy statement, in `run_mad6t.py` or in `mad6t_oneturn.py`?

**Comment by amereghe** [comment]
in the `run_mad6t.py`, i.e. in the executable script in the `sandbox`

File: mad6t_oneturn.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    if 'finished normally' not in outputlines[-2]:
+        print("MADX has not completed properly!")
+        sys.exit(1)
+    else:
+        print("MADX has completed properly!")
+
+    #Check the existence of madx output
+    check_madxout('fc.3', 'fort.3.mad')
+    if os.path.isfile('fc.3.aper'):
+        with open('fort.3.mad', 'w') as fc3:
```
[the remark this answers is not in the document]
Comment (luxiaohan) [reply mad6t_oneturn.py]: Yes, you are right.  The file should be opened with `'a'`.

File: mad6t_oneturn.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        if zp:
+            out_name = dest + filename + '.gz'
+            with open(filename, 'rb') as f_in, gzip.open(out_name, 'wb') as f_out:
+                shutil.copyfileobj(f_in, f_out)
+        else:
+            shutil.copy(filename, dest)
+    else:
+        print("The file %s doesn't exist, download failed!"%filename)
+
+def replace(patterns, replacements, source, dest):
```
[the remark this answers is not in the document]
Comment (luxiaohan) [reply mad6t_oneturn.py]: Actually, I used two `with statement` at first. But sometimes the generated file is empty, I suspect the file isn't flushed normally. So I gave up `with statement`, and declare `close` explicitly. Anyway, I will try it again.

File: study.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        #os.remove('mexe.sh')
+        if len(params) == len(values):
+            for i in range(len(params)):
+                vals[params[i]] = values[i].replace('\n', '')
+        else:
+            print("These should somethings wrong with your input files!")
+        return vals
+
+def peel_str(val):
+    val = val.replace('(', '')
```
[the remark this answers is not in the document]
Comment (luxiaohan) [reply study.py]: Yes, this is more flexible!

File: study.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    val = val.replace(')', '')
+    val = val.replace('\n', '')
+    val = val.replace(' ', '')
+    return val
+
+def num(val):
+    if is_numeral(val):
```
[the remark this answers is not in the document]
Comment (luxiaohan) [reply study.py]: The `isinstance` function is used to check the type of a variable. But for our situation, all the parameters from `*.ini` file are read as `string`.  The function `is_numeral` aim to check if a `string` variable can be converted to a number.

File: mad6t_oneturn.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/python3
+import os
+import re
+import sys
+import copy
+import gzip
+import shutil
+import configparser
+
+def run(args):
```
[the remark this answers is not in the document]
Comment (luxiaohan) [reply mad6t_oneturn.py]: Is the name `run_madx_single_job` for the script or the function `run`?

File: mad6t_oneturn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (amereghe) [reply mad6t_oneturn.py]: uh - strange...

File: study.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (amereghe) [reply study.py]: ah yes, sorry, my bad

**Comment by amereghe** [comment]
Hello @luxiaohan ,
shall I close this pull request?
Cheers,
A.

**Comment by luxiaohan** [comment]
Ok! Just close it.

**Comment by amereghe** [comment]
ok, done.
thanks!

**Pull request closed without merging**